File: 2024/05/5.rb
#!/usr/bin/env ruby

page_deps = Hash.new { |h, k| h[k] = [] }
page_updates = []
$stdin.readlines.each do |line|
  if line.include? '|'
    a, b = line.split('|').map(&:to_i)
    page_deps[a] << b
  elsif line.include? ','
    page_updates << line.split(',').map(&:to_i)
  end
end

total1, total2 = [0, 0]
page_updates.each do |pages|
  sorted_pages = pages.sort do |a, b|
    page_deps[a].include?(b) ? -1 : 1
  end
  middle_page = sorted_pages[pages.size/2]
  if sorted_pages == pages
    total1 += middle_page
  else
    total2 += middle_page
  end
end

puts total1, total2

File: 2024/05/5.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"slices"
	"strconv"
	"strings"
)

func parseInput() (map[int][]int, [][]int) {
	scanner := bufio.NewScanner(os.Stdin)
	pageDeps := map[int][]int{}
	pageUpdates := [][]int{}

	for scanner.Scan() {
		line := scanner.Text()
		if strings.Contains(line, "|") {
			parts := strings.Split(line, "|")
			before, _ := strconv.Atoi(parts[0])
			after, _ := strconv.Atoi(parts[1])
			pageDeps[before] = append(pageDeps[before], after)
		} else if strings.Contains(line, ",") {
			digits := strings.Split(line, ",")
			pages := make([]int, len(digits))
			for i, digit := range digits {
				page, _ := strconv.Atoi(digit)
				pages[i] = page
			}
			pageUpdates = append(pageUpdates, pages)
		}
	}

	return pageDeps, pageUpdates
}

func main() {
	pageDeps, pageUpdates := parseInput()
	total1, total2 := 0, 0

	for _, pages := range pageUpdates {
		sortedPages := make([]int, len(pages))
		copy(sortedPages, pages)

		slices.SortFunc(sortedPages, func(a, b int) int {
			if slices.Contains(pageDeps[a], b) {
				return -1
			}
			return 1
		})

		middlePage := sortedPages[len(sortedPages)/2]

		if slices.Compare(pages, sortedPages) == 0 {
			total1 += middlePage
		} else {
			total2 += middlePage
		}
	}

	fmt.Println(total1)
	fmt.Println(total2)
}
